File: src/crux.ts
/**
 * CrUX (Chrome User Experience Report) API integration.
 *
 * Fetches real-user Core Web Vitals per URL from the CrUX API.
 * Rate limit: 150 requests/minute with an API key, 25/day without.
 * Requires a Google API key with CrUX API enabled.
 *
 * Docs: https://developer.chrome.com/docs/crux/api
 */

export interface CruxData {
  /** Largest Contentful Paint in ms (p75) */
  lcpMs: number | null;
  /** First Input Delay in ms (p75) — deprecated, kept for historical data */
  fidMs: number | null;
  /** Interaction to Next Paint in ms (p75) */
  inpMs: number | null;
  /** Cumulative Layout Shift (p75) */
  cls: number | null;
  /** Time to First Byte in ms (p75) */
  ttfbMs: number | null;
  /** First Contentful Paint in ms (p75) */
  fcpMs: number | null;
  /** Form factor used for the query */
  formFactor: 'PHONE' | 'DESKTOP' | 'ALL';
}

export interface CruxOptions {
  apiKey: string;
  urls: string[];
  formFactor?: 'PHONE' | 'DESKTOP';
  /** Max concurrent requests (default: 10) */
  concurrency?: number;
}

interface CruxMetric {
  percentiles?: { p75: number };
}

interface CruxApiResponse {
  record?: {
    metrics?: {
      largest_contentful_paint?: CruxMetric;
      first_input_delay?: CruxMetric;
      interaction_to_next_paint?: CruxMetric;
      cumulative_layout_shift?: CruxMetric;
      experimental_time_to_first_byte?: CruxMetric;
      first_contentful_paint?: CruxMetric;
    };
  };
  error?: { message: string; code: number };
}

const CRUX_API_URL = 'https://chromeuxreport.googleapis.com/v1/records:queryRecord';

/**
 * Fetch CrUX data for a batch of URLs.
 * Handles rate limiting with a sliding window (150 req/min).
 */
export async function fetchCruxData(options: CruxOptions): Promise<Map<string, CruxData>> {
  const { apiKey, urls, formFactor, concurrency = 10 } = options;
  const results = new Map<string, CruxData>();

  if (!apiKey) {
    throw new Error('CrUX API requires a Google API key (--crux-key). Enable CrUX API at https://console.cloud.google.com');
  }

  // Rate limiting: 150 req/min = ~2.5 req/sec
  const RATE_LIMIT = 150;
  const WINDOW_MS = 60_000;
  const timestamps: number[] = [];

  async function waitForSlot(): Promise<void> {
    while (true) {
      const now = Date.now();
      // Remove timestamps outside the window
      while (timestamps.length > 0 && timestamps[0] < now - WINDOW_MS) {
        timestamps.shift();
      }
      if (timestamps.length < RATE_LIMIT) {
        timestamps.push(now);
        return;
      }
      // Wait until the oldest request exits the window
      const waitMs = timestamps[0] + WINDOW_MS - now + 10;
      await new Promise((resolve) => setTimeout(resolve, waitMs));
    }
  }

  async function fetchSingle(url: string, retries = 0): Promise<void> {
    await waitForSlot();

    const body: Record<string, unknown> = { url };
    if (formFactor) {
      body.formFactor = formFactor;
    }

    try {
      const response = await fetch(`${CRUX_API_URL}?key=${apiKey}`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: JSON.stringify(body),
      });

      if (response.status === 404) {
        // No CrUX data for this URL (not enough traffic)
        return;
      }

      if (!response.ok) {
        const text = await response.text();
        process.stderr.write(`CrUX warning: ${response.status} for ${url}: ${text.substring(0, 200)}\n`);
        return;
      }

      const data = (await response.json()) as CruxApiResponse;

      if (data.error) {
        // API error (e.g., quota exceeded)
        if (data.error.code === 429 && retries < 1) {
          process.stderr.write('CrUX rate limit hit, waiting 60s...\n');
          await new Promise((resolve) => setTimeout(resolve, 60_000));
          return fetchSingle(url, retries + 1);
        }
        return;
      }

      const metrics = data.record?.metrics;
      if (!metrics) return;

      results.set(url, {
        lcpMs: metrics.largest_contentful_paint?.percentiles?.p75 ?? null,
        fidMs: metrics.first_input_delay?.percentiles?.p75 ?? null,
        inpMs: metrics.interaction_to_next_paint?.percentiles?.p75 ?? null,
        cls: metrics.cumulative_layout_shift?.percentiles?.p75 ?? null,
        ttfbMs: metrics.experimental_time_to_first_byte?.percentiles?.p75 ?? null,
        fcpMs: metrics.first_contentful_paint?.percentiles?.p75 ?? null,
        formFactor: formFactor || 'ALL',
      });
    } catch (err) {
      process.stderr.write(`CrUX fetch error for ${url}: ${(err as Error).message}\n`);
    }
  }

  // Process URLs with bounded concurrency
  const queue = [...urls];
  const active: Promise<void>[] = [];

  while (queue.length > 0 || active.length > 0) {
    while (active.length < concurrency && queue.length > 0) {
      const url = queue.shift()!;
      const task = fetchSingle(url).then(() => {
        const idx = active.indexOf(task);
        if (idx !== -1) active.splice(idx, 1);
      });
      active.push(task);
    }
    if (active.length > 0) {
      await Promise.race(active);
    }
  }

  return results;
}

/**
 * Get CrUX assessment for a given metric value.
 * Uses Google's "good", "needs improvement", "poor" thresholds.
 */
export function getCruxAssessment(metric: string, value: number): 'good' | 'needs-improvement' | 'poor' {
  switch (metric) {
    case 'lcp':
      return value <= 2500 ? 'good' : value <= 4000 ? 'needs-improvement' : 'poor';
    case 'fid':
      return value <= 100 ? 'good' : value <= 300 ? 'needs-improvement' : 'poor';
    case 'inp':
      return value <= 200 ? 'good' : value <= 500 ? 'needs-improvement' : 'poor';
    case 'cls':
      return value <= 0.1 ? 'good' : value <= 0.25 ? 'needs-improvement' : 'poor';
    case 'ttfb':
      return value <= 800 ? 'good' : value <= 1800 ? 'needs-improvement' : 'poor';
    case 'fcp':
      return value <= 1800 ? 'good' : value <= 3000 ? 'needs-improvement' : 'poor';
    default:
      return 'needs-improvement';
  }
}

File: internal/integration/crux.go
package integration

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"sync"
	"time"

	"github.com/micelio/micelio/internal/types"
)

var cruxEndpoint = "https://chromeuxreport.googleapis.com/v1/records:queryRecord"

const (
	cruxMaxPerMinute    = 150
	cruxDefaultConc     = 10
	cruxMaxBodySize     = 1 << 20 // 1 MB
	cruxBackoffDuration = 60 * time.Second
)

// setCruxEndpoint overrides the CrUX endpoint (for testing).
func setCruxEndpoint(ep string) { cruxEndpoint = ep }

// CrUX assessment thresholds.
type CrUXAssessment string

const (
	CrUXGood             CrUXAssessment = "good"
	CrUXNeedsImprovement CrUXAssessment = "needs-improvement"
	CrUXPoor             CrUXAssessment = "poor"
)

// CruxClient fetches Chrome UX Report data with rate limiting and concurrency control.
type CruxClient struct {
	httpClient *http.Client
	apiKey     string
	formFactor types.CrUXFormFactor
	timestamps []time.Time
	conc       int
	mu         sync.Mutex
}

// NewCruxClient creates a new CrUX client.
func NewCruxClient(apiKey string, formFactor types.CrUXFormFactor, concurrency int) *CruxClient {
	if concurrency <= 0 {
		concurrency = cruxDefaultConc
	}
	return &CruxClient{
		apiKey:     apiKey,
		formFactor: formFactor,
		conc:       concurrency,
		httpClient: &http.Client{Timeout: 30 * time.Second},
	}
}

// FetchBatch fetches CrUX data for multiple URLs with bounded concurrency.
func (c *CruxClient) FetchBatch(ctx context.Context, urls []string) map[string]*types.CruxData {
	results := make(map[string]*types.CruxData, len(urls))
	var mu sync.Mutex

	sem := make(chan struct{}, c.conc)
	var wg sync.WaitGroup

	for _, u := range urls {
		if ctx.Err() != nil {
			break
		}
		wg.Add(1)
		sem <- struct{}{}
		go func(pageURL string) {
			defer func() { <-sem; wg.Done() }()
			data, err := c.fetch(ctx, pageURL)
			if err != nil {
				return // no data for this URL
			}
			mu.Lock()
			results[pageURL] = data
			mu.Unlock()
		}(u)
	}
	wg.Wait()
	return results
}

func (c *CruxClient) fetch(ctx context.Context, pageURL string) (*types.CruxData, error) {
	if err := c.waitForRateLimit(ctx); err != nil {
		return nil, err
	}

	reqBody := map[string]interface{}{
		"url": pageURL,
	}
	if c.formFactor != "" && c.formFactor != types.FormFactorAll {
		reqBody["formFactor"] = string(c.formFactor)
	}

	bodyBytes, _ := json.Marshal(reqBody)
	endpoint := cruxEndpoint + "?key=" + url.QueryEscape(c.apiKey)

	req, err := http.NewRequestWithContext(ctx, "POST", endpoint, bytes.NewReader(bodyBytes))
	if err != nil {
		return nil, err
	}
	req.Header.Set("Content-Type", "application/json")

	resp, err := c.httpClient.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	respBody, err := io.ReadAll(io.LimitReader(resp.Body, cruxMaxBodySize))
	if err != nil {
		return nil, err
	}

	if resp.StatusCode == http.StatusNotFound {
		return nil, fmt.Errorf("no crux data for %s", pageURL)
	}
	if resp.StatusCode == http.StatusTooManyRequests {
		select {
		case <-time.After(cruxBackoffDuration):
		case <-ctx.Done():
			return nil, ctx.Err()
		}
		return nil, fmt.Errorf("crux: rate limited")
	}
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("crux: HTTP %d", resp.StatusCode)
	}

	return parseCruxResponse(respBody, c.formFactor)
}

func (c *CruxClient) waitForRateLimit(ctx context.Context) error {
	c.mu.Lock()
	defer c.mu.Unlock()

	now := time.Now()
	cutoff := now.Add(-time.Minute)

	// Trim old timestamps
	valid := 0
	for _, ts := range c.timestamps {
		if ts.After(cutoff) {
			c.timestamps[valid] = ts
			valid++
		}
	}
	c.timestamps = c.timestamps[:valid]

	if len(c.timestamps) >= cruxMaxPerMinute {
		sleepUntil := c.timestamps[0].Add(time.Minute)
		c.mu.Unlock()
		select {
		case <-time.After(time.Until(sleepUntil)):
		case <-ctx.Done():
			c.mu.Lock()
			return ctx.Err()
		}
		c.mu.Lock()
		// Re-trim after sleep
		now = time.Now()
		cutoff = now.Add(-time.Minute)
		valid = 0
		for _, ts := range c.timestamps {
			if ts.After(cutoff) {
				c.timestamps[valid] = ts
				valid++
			}
		}
		c.timestamps = c.timestamps[:valid]
	}

	c.timestamps = append(c.timestamps, time.Now())
	return nil
}

func parseCruxResponse(body []byte, formFactor types.CrUXFormFactor) (*types.CruxData, error) {
	var raw struct {
		Record struct {
			Metrics map[string]struct {
				Percentiles struct {
					P75 json.RawMessage `json:"p75"`
				} `json:"percentiles"`
			} `json:"metrics"`
		} `json:"record"`
	}

	if err := json.Unmarshal(body, &raw); err != nil {
		return nil, fmt.Errorf("crux: parse: %w", err)
	}

	data := &types.CruxData{FormFactor: formFactor}
	if formFactor == "" {
		data.FormFactor = types.FormFactorAll
	}

	metrics := raw.Record.Metrics
	data.LCPMs = extractP75Float(metrics, "largest_contentful_paint")
	data.FIDMs = extractP75Float(metrics, "first_input_delay")
	data.INPMs = extractP75Float(metrics, "interaction_to_next_paint")
	data.CLS = extractP75Float(metrics, "cumulative_layout_shift")
	data.TTFBMs = extractP75Float(metrics, "experimental_time_to_first_byte")
	data.FCPMs = extractP75Float(metrics, "first_contentful_paint")

	return data, nil
}

func extractP75Float(metrics map[string]struct {
	Percentiles struct {
		P75 json.RawMessage `json:"p75"`
	} `json:"percentiles"`
}, key string) *float64 {
	m, ok := metrics[key]
	if !ok || m.Percentiles.P75 == nil {
		return nil
	}
	var v float64
	if err := json.Unmarshal(m.Percentiles.P75, &v); err != nil {
		return nil
	}
	return &v
}

// GetCrUXAssessment evaluates a metric value against Google's CWV thresholds.
func GetCrUXAssessment(metric string, value float64) CrUXAssessment {
	thresholds := map[string][2]float64{
		"lcp":  {2500, 4000},
		"fid":  {100, 300},
		"inp":  {200, 500},
		"cls":  {0.1, 0.25},
		"ttfb": {800, 1800},
		"fcp":  {1800, 3000},
	}
	t, ok := thresholds[metric]
	if !ok {
		return CrUXGood
	}
	if value <= t[0] {
		return CrUXGood
	}
	if value <= t[1] {
		return CrUXNeedsImprovement
	}
	return CrUXPoor
}
